File: app/src/main/java/com/trevorwiebe/trackacow/presentation/activities/DrugsGivenReportActivity.java
package com.trevorwiebe.trackacow.presentation.activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.TextView;

import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.domain.adapters.TimeDrugRvAdapter;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.drug.QueryAllDrugs;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.archivedLot.QueryArchivedLotsByLotId;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.drugsGiven.QueryDrugsGivenByLotIdAndDateRange;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.lot.QueryLotByLotId;
import com.trevorwiebe.trackacow.data.entities.ArchivedLotEntity;
import com.trevorwiebe.trackacow.data.entities.DrugEntity;
import com.trevorwiebe.trackacow.data.entities.DrugsGivenEntity;
import com.trevorwiebe.trackacow.data.entities.LotEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.Utility;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.concurrent.TimeUnit;

public class DrugsGivenReportActivity extends AppCompatActivity implements
        QueryAllDrugs.OnAllDrugsLoaded,
        QueryDrugsGivenByLotIdAndDateRange.OnDrugsGivenByLotIdAndDateRangeLoaded,
        QueryLotByLotId.OnLotByLotIdLoaded,
        QueryArchivedLotsByLotId.OnArchivedLotLoaded{

    private static final String TAG = "DrugsGivenReportActivit";

    private String mLotId;
    private Integer mReportType;
    private final HashMap<String, String> mDrugKeyAndName = new HashMap<>();

    private final Calendar mStartCalendar = Calendar.getInstance();
    private final Calendar mEndCalendar = Calendar.getInstance();

    private int mReportDateRangeType = 0;

    private DatePickerDialog.OnDateSetListener mStartDatePicker;
    private DatePickerDialog.OnDateSetListener mEndDatePicker;

    private TextView mNoDrugsGivenTv;
    private RecyclerView mDayRv;
    private TimeDrugRvAdapter timeDrugRvAdapter;
    private Button mStartDateButton;
    private Button mEndDateButton;
    private Button mQuickYesterday;
    private Button mQuickMonth;
    private Button mQuickAll;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_drugs_given_report);

        mStartDateButton = findViewById(R.id.start_date_btn);
        mEndDateButton = findViewById(R.id.end_date_btn);
        mQuickYesterday = findViewById(R.id.quick_yesterday);
        mQuickMonth = findViewById(R.id.quick_month);
        mQuickAll = findViewById(R.id.quick_all);

        mNoDrugsGivenTv = findViewById(R.id.no_drugs_given_report_tv);
        mDayRv = findViewById(R.id.drug_report_rv);
        mDayRv.setLayoutManager(new LinearLayoutManager(this));
        timeDrugRvAdapter = new TimeDrugRvAdapter(null);
        mDayRv.setAdapter(timeDrugRvAdapter);

        Intent intent = getIntent();
        mLotId = intent.getStringExtra("lotId");
        mReportType = intent.getIntExtra("reportType", -1);

        if(savedInstanceState != null){

            long startDate = savedInstanceState.getLong("startLong");
            long endDate = savedInstanceState.getLong("endLong");

            setSelectedButton(savedInstanceState.getInt("reportType"));

            mStartCalendar.setTimeInMillis(startDate);
            mEndCalendar.setTimeInMillis(endDate);

        }else {

            long millisInADay = TimeUnit.DAYS.toMillis(1);

            long yesterday_milliseconds = System.currentTimeMillis() - millisInADay;

            mStartCalendar.setTimeInMillis(yesterday_milliseconds);
            mStartCalendar.set(Calendar.HOUR, 0);
            mStartCalendar.set(Calendar.MINUTE, 0);
            mStartCalendar.set(Calendar.SECOND, 0);
            mStartCalendar.set(Calendar.MILLISECOND, 0);

            mEndCalendar.set(Calendar.HOUR, mEndCalendar.getMaximum(Calendar.HOUR));
            mEndCalendar.set(Calendar.MINUTE, mEndCalendar.getMaximum(Calendar.MINUTE));
            mEndCalendar.set(Calendar.SECOND, mEndCalendar.getMaximum(Calendar.SECOND));
            mEndCalendar.set(Calendar.MILLISECOND, mEndCalendar.getMaximum(Calendar.MILLISECOND));

            setSelectedButton(Constants.YESTERDAY);
        }

        mStartDateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                new DatePickerDialog(DrugsGivenReportActivity.this,
                        mStartDatePicker,
                        mStartCalendar.get(Calendar.YEAR),
                        mStartCalendar.get(Calendar.MONTH),
                        mStartCalendar.get(Calendar.DAY_OF_MONTH))
                        .show();
            }
        });

        mStartDatePicker = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                mStartCalendar.set(Calendar.YEAR, year);
                mStartCalendar.set(Calendar.MONTH, month);
                mStartCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                mStartDateButton.setText(Utility.convertMillisToDate(mStartCalendar.getTimeInMillis()));

                long selectedStartDate = mStartCalendar.getTimeInMillis();
                long selectedEndDate = mEndCalendar.getTimeInMillis();

                setSelectedButton(Constants.CUSTOM);

                new QueryDrugsGivenByLotIdAndDateRange(DrugsGivenReportActivity.this, mLotId, selectedStartDate, selectedEndDate).execute(DrugsGivenReportActivity.this);
            }
        };

        mEndDateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new DatePickerDialog(DrugsGivenReportActivity.this,
                        mEndDatePicker,
                        mEndCalendar.get(Calendar.YEAR),
                        mEndCalendar.get(Calendar.MONTH),
                        mEndCalendar.get(Calendar.DAY_OF_MONTH))
                        .show();
            }
        });

        mEndDatePicker = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                mEndCalendar.set(Calendar.YEAR, year);
                mEndCalendar.set(Calendar.MONTH, month);
                mEndCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                mEndDateButton.setText(Utility.convertMillisToDate(mEndCalendar.getTimeInMillis()));

                long selectedStartDate = mStartCalendar.getTimeInMillis();
                long selectedEndDate = mEndCalendar.getTimeInMillis();

                setSelectedButton(Constants.CUSTOM);

                new QueryDrugsGivenByLotIdAndDateRange(DrugsGivenReportActivity.this, mLotId, selectedStartDate, selectedEndDate).execute(DrugsGivenReportActivity.this);
            }
        };

        mQuickAll.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(mReportType == Constants.LOT) {
                    new QueryLotByLotId(mLotId, DrugsGivenReportActivity.this).execute(DrugsGivenReportActivity.this);
                }else{
                    new QueryArchivedLotsByLotId(mLotId, DrugsGivenReportActivity.this).execute(DrugsGivenReportActivity.this);
                }

                setSelectedButton(Constants.ALL);
            }
        });

        mQuickMonth.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                mStartCalendar.setTimeInMillis(System.currentTimeMillis());
                mStartCalendar.add(Calendar.MONTH, -1);
                mEndCalendar.setTimeInMillis(System.currentTimeMillis());

                long startLong = mStartCalendar.getTimeInMillis();
                long endLong = mEndCalendar.getTimeInMillis();

                String friendlyStartDate = Utility.convertMillisToDate(startLong);
                String friendlyEndDate = Utility.convertMillisToDate(endLong);

                mStartDateButton.setText(friendlyStartDate);
                mEndDateButton.setText(friendlyEndDate);

                new QueryDrugsGivenByLotIdAndDateRange(DrugsGivenReportActivity.this,
                        mLotId, startLong, endLong)
                        .execute(DrugsGivenReportActivity.this);

                setSelectedButton(Constants.MONTH);
            }
        });

        mQuickYesterday.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                long millisInADay = 86400000;

                long today = System.currentTimeMillis();
                long yesterday = today - millisInADay;

                mStartCalendar.setTimeInMillis(yesterday);
                mStartCalendar.set(Calendar.HOUR, 0);
                mStartCalendar.set(Calendar.MINUTE, 0);
                mStartCalendar.set(Calendar.SECOND, 0);

                mEndCalendar.setTimeInMillis(System.currentTimeMillis());

                long startLong = mStartCalendar.getTimeInMillis();
                long endLong = mEndCalendar.getTimeInMillis();

                String friendlyStartDate = Utility.convertMillisToDate(startLong);
                String friendlyEndDate = Utility.convertMillisToDate(endLong);

                mStartDateButton.setText(friendlyStartDate);
                mEndDateButton.setText(friendlyEndDate);

                new QueryDrugsGivenByLotIdAndDateRange(DrugsGivenReportActivity.this,
                        mLotId, startLong, endLong)
                        .execute(DrugsGivenReportActivity.this);

                setSelectedButton(Constants.YESTERDAY);

            }
        });

        new QueryAllDrugs(this).execute(this);
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putLong("startLong", mStartCalendar.getTimeInMillis());
        outState.putLong("endLong", mEndCalendar.getTimeInMillis());
        outState.putInt("reportType", mReportDateRangeType);
        super.onSaveInstanceState(outState);
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return false;
    }

    @Override
    public void onAllDrugsLoaded(ArrayList<DrugEntity> drugEntities) {

        for (int a = 0; a < drugEntities.size(); a++) {
            DrugEntity drugEntity = drugEntities.get(a);
            String drugId = drugEntity.getDrugCloudDatabaseId();
            String drugName = drugEntity.getDrugName();
            mDrugKeyAndName.put(drugId, drugName);
        }

        long startLong = mStartCalendar.getTimeInMillis();
        long endLong = mEndCalendar.getTimeInMillis();

        String friendlyStartDate = Utility.convertMillisToDate(startLong);
        String friendlyEndDate = Utility.convertMillisToDate(endLong);

        mStartDateButton.setText(friendlyStartDate);
        mEndDateButton.setText(friendlyEndDate);

        new QueryDrugsGivenByLotIdAndDateRange(this, mLotId, startLong, endLong).execute(this);

    }

    @Override
    public void onDrugsGivenByLotIdAndDateRangeLoaded(ArrayList<DrugsGivenEntity> drugsGivenEntities) {
        ArrayList<DrugsGivenEntity> drugsGivenListCondensed = new ArrayList<>();

        for (int e = 0; e < drugsGivenEntities.size(); e++) {
            DrugsGivenEntity drugsGivenEntity = drugsGivenEntities.get(e);
            String drugId = drugsGivenEntity.getDrugId();
            int amountGiven = drugsGivenEntity.getAmountGiven();
            if (findAndUpdateDrugReports(drugId, amountGiven, drugsGivenListCondensed) == 0) {
                drugsGivenListCondensed.add(drugsGivenEntity);
            }
        }

        for (int f = 0; f < drugsGivenListCondensed.size(); f++) {
            DrugsGivenEntity drugGivenList = drugsGivenListCondensed.get(f);
            String drugId = drugGivenList.getDrugId();
            String drugName = mDrugKeyAndName.get(drugId);
            drugGivenList.setDrugId(drugName);
            drugsGivenListCondensed.remove(f);
            drugsGivenListCondensed.add(f, drugGivenList);
        }

        if (drugsGivenListCondensed.size() == 0) {
            mNoDrugsGivenTv.setVisibility(View.VISIBLE);
            mDayRv.setVisibility(View.GONE);
        } else {
            mNoDrugsGivenTv.setVisibility(View.GONE);
            mDayRv.setVisibility(View.VISIBLE);
            timeDrugRvAdapter.swapData(drugsGivenListCondensed);
        }
    }

    private int findAndUpdateDrugReports(String drugId, int amountGiven, ArrayList<DrugsGivenEntity> drugReportsObjects) {
        for (int r = 0; r < drugReportsObjects.size(); r++) {
            DrugsGivenEntity drugReportsObject = drugReportsObjects.get(r);
            if (drugReportsObject.getDrugId().endsWith(drugId)) {
                int currentAmount = drugReportsObject.getAmountGiven();
                int amountToUpdateTo = currentAmount + amountGiven;
                drugReportsObject.setAmountGiven(amountToUpdateTo);
                drugReportsObjects.remove(r);
                drugReportsObjects.add(r, drugReportsObject);
                return 1;
            }
        }
        return 0;
    }

    @Override
    public void onLotByLotIdLoaded(LotEntity lotEntity) {
        long start_date = lotEntity.getDate();
        long end_date = System.currentTimeMillis();

        String friendlyStartDate = Utility.convertMillisToDate(start_date);
        String friendlyEndDate = Utility.convertMillisToDate(end_date);

        mStartCalendar.setTimeInMillis(start_date);
        mEndCalendar.setTimeInMillis(end_date);

        mStartDateButton.setText(friendlyStartDate);
        mEndDateButton.setText(friendlyEndDate);

        new QueryDrugsGivenByLotIdAndDateRange(this, mLotId, start_date, end_date).execute(this);
    }

    @Override
    public void onArchivedLotLoaded(ArchivedLotEntity archivedLotEntity) {
        long start_date = archivedLotEntity.getDateStarted();
        long end_date = archivedLotEntity.getDateEnded();

        String friendlyStartDate = Utility.convertMillisToDate(start_date);
        String friendlyEndDate = Utility.convertMillisToDate(end_date);

        mStartCalendar.setTimeInMillis(start_date);
        mEndCalendar.setTimeInMillis(end_date);

        mStartDateButton.setText(friendlyStartDate);
        mEndDateButton.setText(friendlyEndDate);

        new QueryDrugsGivenByLotIdAndDateRange(this, mLotId, start_date, end_date).execute(this);
    }

    private void setSelectedButton(int reportType){

        mReportDateRangeType = reportType;

        switch (reportType){
            case Constants.YESTERDAY:
                mQuickYesterday.setBackground(getDrawable(R.drawable.accent_sign_in_btn));
                mQuickYesterday.setTextColor(getResources().getColor(android.R.color.white));

                mQuickMonth.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickMonth.setTextColor(getResources().getColor(R.color.colorAccent));

                mQuickAll.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickAll.setTextColor(getResources().getColor(R.color.colorAccent));
                break;
            case Constants.MONTH:
                mQuickYesterday.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickYesterday.setTextColor(getResources().getColor(R.color.colorAccent));

                mQuickMonth.setBackground(getDrawable(R.drawable.accent_sign_in_btn));
                mQuickMonth.setTextColor(getResources().getColor(android.R.color.white));

                mQuickAll.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickAll.setTextColor(getResources().getColor(R.color.colorAccent));
                break;
            case Constants.ALL:
                mQuickYesterday.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickYesterday.setTextColor(getResources().getColor(R.color.colorAccent));

                mQuickAll.setBackground(getDrawable(R.drawable.accent_sign_in_btn));
                mQuickAll.setTextColor(getResources().getColor(android.R.color.white));

                mQuickMonth.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickMonth.setTextColor(getResources().getColor(R.color.colorAccent));
                break;
            case Constants.CUSTOM:
                mQuickYesterday.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickYesterday.setTextColor(getResources().getColor(R.color.colorAccent));

                mQuickAll.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickAll.setTextColor(getResources().getColor(R.color.colorAccent));

                mQuickMonth.setBackground(getDrawable(R.drawable.white_sign_in_btn_accent_border));
                mQuickMonth.setTextColor(getResources().getColor(R.color.colorAccent));
                break;
        }
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/drugs_given_reports/DrugsGivenReportActivity.kt
package com.trevorwiebe.trackacow.presentation.drugs_given_reports

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.os.Build.VERSION
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.trevorwiebe.trackacow.R
import com.trevorwiebe.trackacow.domain.models.lot.LotModel
import com.trevorwiebe.trackacow.domain.utils.Constants
import com.trevorwiebe.trackacow.domain.utils.DataSource
import com.trevorwiebe.trackacow.domain.utils.Utility
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.util.*

@AndroidEntryPoint
class DrugsGivenReportActivity : AppCompatActivity() {

    private lateinit var timeDrugRvAdapter: TimeDrugRvAdapter
    private var mLotModel: LotModel? = null
    private val mStartCalendar = Calendar.getInstance()
    private val mEndCalendar = Calendar.getInstance()

    private lateinit var mStartDatePicker: OnDateSetListener
    private lateinit var mEndDatePicker: OnDateSetListener
    private lateinit var mNoDrugsGivenTv: TextView
    private lateinit var mDrugsReportRV: RecyclerView
    private lateinit var mProgressIndicator: LinearProgressIndicator
    private lateinit var mStartDateButton: Button
    private lateinit var mEndDateButton: Button
    private lateinit var mLast24HRs: Button
    private lateinit var mQuickMonth: Button
    private lateinit var mQuickAll: Button

    private val drugsGivenViewModel: DrugsGivenViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_drugs_given_report)

        mStartDateButton = findViewById(R.id.start_date_btn)
        mEndDateButton = findViewById(R.id.end_date_btn)
        mLast24HRs = findViewById(R.id.quick_yesterday)
        mQuickMonth = findViewById(R.id.quick_month)
        mQuickAll = findViewById(R.id.quick_all)
        mNoDrugsGivenTv = findViewById(R.id.no_drugs_given_report_tv)
        mProgressIndicator = findViewById(R.id.drugs_given_report_progress_indicator)
        mDrugsReportRV = findViewById(R.id.drug_report_rv)
        mDrugsReportRV.layoutManager = LinearLayoutManager(this)
        timeDrugRvAdapter = TimeDrugRvAdapter()
        mDrugsReportRV.adapter = timeDrugRvAdapter

        @Suppress("DEPRECATION")
        mLotModel = if (VERSION.SDK_INT >= 33) {
            intent.getParcelableExtra("lotModel", LotModel::class.java)
        } else {
            intent.getParcelableExtra("lotModel")
        }

        title = resources.getString(R.string.drug_report_title, mLotModel?.lotName)

        if (savedInstanceState != null) {
            val startDate = savedInstanceState.getLong("startLong")
            val endDate = savedInstanceState.getLong("endLong")
            setSelectedButton(savedInstanceState.getInt("reportType"))
            mStartCalendar.timeInMillis = startDate
            mEndCalendar.timeInMillis = endDate
        } else {
            val millisInADay: Long = 86400000
            val today = System.currentTimeMillis()
            val yesterday = today - millisInADay
            mStartCalendar.timeInMillis = yesterday
            mEndCalendar.timeInMillis = System.currentTimeMillis()
            val startLong = mStartCalendar.timeInMillis
            val endLong = mEndCalendar.timeInMillis
            mStartDateButton.text = Utility.convertMillisToDate(startLong)
            mEndDateButton.text = Utility.convertMillisToDate(endLong)

            drugsGivenViewModel.onEvent(
                DrugsGivenUiEvent.CustomDateSelected(
                    mLotModel?.lotCloudDatabaseId ?: "", startLong, endLong
                )
            )

            setSelectedButton(Constants.YESTERDAY)
        }

        mStartDateButton.text = Utility.convertMillisToDate(mStartCalendar.timeInMillis)
        mEndDateButton.text = Utility.convertMillisToDate(mEndCalendar.timeInMillis)

        mStartDateButton.setOnClickListener {
            DatePickerDialog(
                this@DrugsGivenReportActivity,
                mStartDatePicker,
                mStartCalendar[Calendar.YEAR],
                mStartCalendar[Calendar.MONTH],
                mStartCalendar[Calendar.DAY_OF_MONTH]
            )
                .show()
        }

        mStartDatePicker = OnDateSetListener { view, year, month, dayOfMonth ->
            mStartCalendar[Calendar.YEAR] = year
            mStartCalendar[Calendar.MONTH] = month
            mStartCalendar[Calendar.DAY_OF_MONTH] = dayOfMonth
            mStartDateButton.text = Utility.convertMillisToDate(mStartCalendar.timeInMillis)
            val selectedStartDate = mStartCalendar.timeInMillis
            val selectedEndDate = mEndCalendar.timeInMillis
            setSelectedButton(Constants.CUSTOM)

            drugsGivenViewModel.onEvent(
                DrugsGivenUiEvent.CustomDateSelected(
                    mLotModel?.lotCloudDatabaseId ?: "", selectedStartDate, selectedEndDate
                )
            )

        }

        mEndDateButton.setOnClickListener {
            DatePickerDialog(
                this@DrugsGivenReportActivity,
                mEndDatePicker,
                mEndCalendar[Calendar.YEAR],
                mEndCalendar[Calendar.MONTH],
                mEndCalendar[Calendar.DAY_OF_MONTH]
            )
                .show()
        }

        mEndDatePicker = OnDateSetListener { view, year, month, dayOfMonth ->
            mEndCalendar[Calendar.YEAR] = year
            mEndCalendar[Calendar.MONTH] = month
            mEndCalendar[Calendar.DAY_OF_MONTH] = dayOfMonth
            mEndDateButton.text = Utility.convertMillisToDate(mEndCalendar.timeInMillis)
            val selectedStartDate = mStartCalendar.timeInMillis
            val selectedEndDate = mEndCalendar.timeInMillis
            setSelectedButton(Constants.CUSTOM)

            drugsGivenViewModel.onEvent(
                DrugsGivenUiEvent.CustomDateSelected(
                    mLotModel?.lotCloudDatabaseId ?: "", selectedStartDate, selectedEndDate
                )
            )
        }

        mQuickAll.setOnClickListener {
            if (mLotModel != null) {
                val lotId = mLotModel!!.lotCloudDatabaseId

                val lotStart = mLotModel!!.date
                val endDate = System.currentTimeMillis()

                mStartCalendar.timeInMillis = lotStart
                mEndCalendar.timeInMillis = endDate
                val friendlyStartDate = Utility.convertMillisToDate(lotStart)
                val friendlyEndDate = Utility.convertMillisToDate(endDate)
                mStartDateButton.text = friendlyStartDate
                mEndDateButton.text = friendlyEndDate

                drugsGivenViewModel.onEvent(
                    DrugsGivenUiEvent.CustomDateSelected(
                        lotId,
                        lotStart,
                        endDate
                    )
                )
            } else {
                // Archive Lot
            }
            setSelectedButton(Constants.ALL)
        }

        mQuickMonth.setOnClickListener {
            mStartCalendar.timeInMillis = System.currentTimeMillis()
            mStartCalendar.add(Calendar.MONTH, -1)
            mEndCalendar.timeInMillis = System.currentTimeMillis()
            val startLong = mStartCalendar.timeInMillis
            val endLong = mEndCalendar.timeInMillis
            val friendlyStartDate = Utility.convertMillisToDate(startLong)
            val friendlyEndDate = Utility.convertMillisToDate(endLong)
            mStartDateButton.text = friendlyStartDate
            mEndDateButton.text = friendlyEndDate

            drugsGivenViewModel.onEvent(
                DrugsGivenUiEvent.CustomDateSelected(
                    mLotModel?.lotCloudDatabaseId ?: "", startLong, endLong
                )
            )

            setSelectedButton(Constants.MONTH)
        }

        mLast24HRs.setOnClickListener {
            val millisInADay: Long = 86400000
            val today = System.currentTimeMillis()
            val yesterday = today - millisInADay
            mStartCalendar.timeInMillis = yesterday
            mEndCalendar.timeInMillis = System.currentTimeMillis()
            val startLong = mStartCalendar.timeInMillis
            val endLong = mEndCalendar.timeInMillis
            mStartDateButton.text = Utility.convertMillisToDate(startLong)
            mEndDateButton.text = Utility.convertMillisToDate(endLong)

            drugsGivenViewModel.onEvent(
                DrugsGivenUiEvent.CustomDateSelected(
                    mLotModel?.lotCloudDatabaseId ?: "", startLong, endLong
                )
            )

            setSelectedButton(Constants.YESTERDAY)
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.RESUMED) {
                drugsGivenViewModel.uiState.collect { drugsGivenUiState ->

                    if (drugsGivenUiState.isFetchingFromCloud && drugsGivenUiState.dataSource == DataSource.Local) {
                        mProgressIndicator.visibility = View.VISIBLE
                    } else {
                        mProgressIndicator.visibility = View.INVISIBLE
                    }

                    if (drugsGivenUiState.drugsGivenAndDrugList.isEmpty()) {
                        mNoDrugsGivenTv.visibility = View.VISIBLE
                        mDrugsReportRV.visibility = View.GONE
                    } else {
                        mNoDrugsGivenTv.visibility = View.GONE
                        mDrugsReportRV.visibility = View.VISIBLE
                        timeDrugRvAdapter.swapData(drugsGivenUiState.drugsGivenAndDrugList)
                    }
                }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putLong("startLong", mStartCalendar.timeInMillis)
        outState.putLong("endLong", mEndCalendar.timeInMillis)
        super.onSaveInstanceState(outState)
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return false
    }

    private fun setSelectedButton(reportType: Int) {
        when (reportType) {
            Constants.YESTERDAY -> {
                mLast24HRs.background =
                    AppCompatResources.getDrawable(this, R.drawable.accent_sign_in_btn)
                mLast24HRs.setTextColor(ContextCompat.getColor(this, android.R.color.white))
                mQuickMonth.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mQuickMonth.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
                mQuickAll.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mQuickAll.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
            }
            Constants.MONTH -> {
                mLast24HRs.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mLast24HRs.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
                mQuickMonth.background =
                    AppCompatResources.getDrawable(this, R.drawable.accent_sign_in_btn)
                mQuickMonth.setTextColor(ContextCompat.getColor(this, android.R.color.white))
                mQuickAll.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mQuickAll.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
            }
            Constants.ALL -> {
                mLast24HRs.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mLast24HRs.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
                mQuickAll.background =
                    AppCompatResources.getDrawable(this, R.drawable.accent_sign_in_btn)
                mQuickAll.setTextColor(ContextCompat.getColor(this, android.R.color.white))
                mQuickMonth.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mQuickMonth.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
            }
            Constants.CUSTOM -> {
                mLast24HRs.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mLast24HRs.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
                mQuickAll.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mQuickAll.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
                mQuickMonth.background =
                    AppCompatResources.getDrawable(this, R.drawable.white_sign_in_btn_accent_border)
                mQuickMonth.setTextColor(ContextCompat.getColor(this, R.color.colorAccent))
            }
        }
    }
}